### ISSUE TITLE
Proposal to add iOS VoiceOver Two-finger scrub 

### ISSUE THREAD

**Issue by yasargil** [flutter/flutter#74246]
There is a VoiceOver gesture to navigate to the previous screen (Two-finger scrub). It looks like flutter is not implementing it. See documentation on https://support.apple.com/guide/iphone/learn-voiceover-gestures-iph3e2e2281/14.0/ios/14.0

**Comment by robertguggenmos** [comment]
Any news update that? We also wait for the support of the two finger scrub z gesture in flutter. We got some feedback from a blind person, who told us that this functionality is quite important in VoiceOver mode.

**Comment by guenterhanke** [comment]
Hi,
i am blind myself and use my iPhone with VoiceOver.
Without this gesture i'll have to search for the back button on the display with my finger manually.
Then I and have tto doubletab this button to go back.
This takes me several seconds.

 This is a very big problem!
Even if i have to go back more times.

With this gesture implemented i only would perform a z with 2 fingers on my display.

Thank you for implementing it
Guenter from Germany.

**Comment by yasargil** [comment]
A simple solution would be subclassing the FlutterViewController and adding following method:

```
- (BOOL)accessibilityPerformEscape {
    [self popRoute];
    return YES;
}
```

And in dart/flutter code it should work out of the box like the android back button.

**Comment by wapscccs** [comment]
Hello, I would like to ask where I need to rewrite FlutterViewController? Is there a demo for reference? 
@yasargil

**Comment by yasargil** [comment]
Copy these classes in to your xcodeworkspace and modify Main.storyboard

`<viewController id="BYZ-38-t0r" customClass="OVPFlutterViewController" sceneMemberID="viewController">`

[Archiv.zip](https://github.com/flutter/flutter/files/7755852/Archiv.zip)

**Comment by wapscccs** [comment]
Hello, this is my Demo code. According to your method, it still doesn't seem to work. Could it be convenient for you to take a look? Thank you very much
[demo.zip](https://github.com/flutter/flutter/files/7756304/demo.zip)

**Comment by yasargil** [comment]
@wapscccs i added the class and modified the storyboard.  
[demo.zip](https://github.com/flutter/flutter/files/7756652/demo.zip)

**Comment by wapscccs** [comment]
@yasargil Thank you very much, you helped me a lot~

**Comment by luomo-pro** [comment]
hi
@yasargil
Thank you for the solution.
Although this can solve this problem, I think it should be supported by flutter by default.
After all, you cannot ask every developer to make this modification.
And this gesture is very important for blind users, and it makes sense to support it by default.
In iOS native, this gesture is supported by default.
I also submitted an issue before
#95567
I hope that flutter can be supported by default as soon as possible, so that countless software that uses flutter will also support it by default.

**Comment by wapscccs** [comment]
> @wapscccs i added the class and modified the storyboard. [demo.zip](https://github.com/flutter/flutter/files/7756652/demo.zip)

@yasargil  Hello, I want to ask you again,
After running the project according to your method, I found that the third-party libraries I installed did not work, such as Flutter_tts,
Since I don't know much about OC, I don't know how to deal with this?
Or can it be added in GeneratedPluginRegistrant? 
Thank you so much

**Comment by yasargil** [comment]
@wapscccs i don't use that plugin, but others and they are still working. My change should also not touch any plugin code. So i am not sure why it would stop working.

**Comment by wapscccs** [comment]
@yasargil  Okay, thank you very much, but I think any program will use third-party plug-ins. Once the third-party plug-ins are referenced, the third-party plug-ins will not work properly if you use this method. So I wonder if it can be added to the default ios project file generated by Flutter instead of adding two files separately

**Comment by yasargil** [comment]
> Once the third-party plug-ins are referenced, the third-party plug-ins will not work properly if you use this method.
That is not what we do observer in our project, we have multiple plugins active and working. I' am sorry but i can't provide you more support at the moment.

**Comment by wapscccs** [comment]
> > Once the third-party plug-ins are referenced, the third-party plug-ins will not work properly if you use this method.
> 
> That is not what we do observer in our project, we have multiple plugins active and working. I' am sorry but i can't provide you more support at the moment.

Okay thank you

**Comment by wapscccs** [comment]
@yasargil The problem is solved, my code was a bit problem before, thank you again~

**Comment by vintage** [comment]
Any news about the two-finger scrub gesture? Currently working at adjusting my game for blind players. Burnt several hours adding `Semantics`/`MergeSemantics`/`BlockSemantics` into the widget tree, as well as making extra adjustments "visible" only for blind peeps. Unfortunately the biggest issue I'm trying to handle right now is the VO back gesture which is not exposed in any way by the framework.

**Comment by hellohuanlin** [comment]
> 

It seems reasonable to just move this logic to [FlutterViewController.mm](https://github.com/flutter/engine/blob/main/shell/platform/darwin/ios/framework/Source/FlutterViewController.mm) directly, which should achieve the same effect as subclassing it.

**Comment by vsomayaji** [comment]
Scratch my earlier [workaround](https://github.com/flutter/flutter/issues/74246#issuecomment-1354021630). While it does work, it impacts VoiceOver's ability to explore the screen. 

We're now borrowing an idea from https://github.com/flutter/flutter/issues/3816#issuecomment-659922085 and adding a custom `BackButton` to our app bar that uses [`SemanticsProperties.onDismiss`](https://api.flutter.dev/flutter/semantics/SemanticsProperties/onDismiss.html). Updated code sample below.

<details><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';
import 'package:go_router/go_router.dart';

void main() {
  WidgetsFlutterBinding.ensureInitialized();

  // Auto-enable accessibility for our Blind and Low Vision customers.
  RendererBinding.instance.setSemanticsEnabled(true);

  runApp(const MyApp());
}

final GoRouter _router = GoRouter(
  routes: <RouteBase>[
    GoRoute(
      path: '/',
      builder: (BuildContext context, GoRouterState state) {
        return const HomeScreen();
      },
      routes: <RouteBase>[
        GoRoute(
          path: 'details',
          builder: (BuildContext context, GoRouterState state) {
            return const DetailsScreen();
          },
        ),
      ],
    ),
  ],
);

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp.router(
      routerConfig: _router,
    );
  }
}

class HomeScreen extends StatelessWidget {
  const HomeScreen({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Home Screen'),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            ElevatedButton(
              onPressed: () => context.go('/details'),
              child: const Text('Go to the Details screen'),
            ),
          ],
        ),
      ),
    );
  }
}

class DetailsScreen extends StatelessWidget {
  const DetailsScreen({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        leading: DismissibleBackButton(
          onDismiss: () => context.go('/'),
        ),
        title: const Text('Details Screen'),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <ElevatedButton>[
            ElevatedButton(
              onPressed: () => context.go('/'),
              child: const Text('Go back to the Home screen'),
            ),
          ],
        ),
      ),
    );
  }
}

/// A custom [BackButton] that supports the VoiceOver two-finger scrub gesture.
///
/// This workaround for https://github.com/flutter/flutter/issues/74246 was borrowed from
/// https://github.com/flutter/flutter/issues/3816#issuecomment-659922085.
class DismissibleBackButton extends StatelessWidget {
  const DismissibleBackButton({Key? key, this.onDismiss}) : super(key: key);

  final VoidCallback? onDismiss;

  @override
  Widget build(BuildContext context) {
    final ModalRoute<dynamic>? parentRoute = ModalRoute.of(context);
    final bool canPop = parentRoute?.canPop ?? false;

    // If we can't pop the current route, don't show a button.
    if (!canPop) {
      return const SizedBox.shrink();
    }

    final VoidCallback? onPressed = onDismiss ?? () => Navigator.of(context).maybePop();

    return Semantics(
      button: true,
      excludeSemantics: true,
      label: MaterialLocalizations.of(context).backButtonTooltip,
      onDismiss: onPressed,
      child: BackButton(
        onPressed: onPressed,
      ),
    );
  }
}
```
</details>
